power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -583,6 +583,27 @@ template <symmetry_tag sym> struct NodeState {
     }
 };
 
+class RankIterator {
+  public:
+    RankIterator() = default;
+
+    void iterate_ranks(auto const& regulator_order, auto adjust_transformer_in_rank, bool& tap_changed,
+                       std::vector<IntS>& iterations_per_rank, Idx& rank_index) {
```

Comment:
do you think it makes more sense to make `iterations_per_rank` and `rank_index` member variables (private + public `constexpr` getters for testability)? The `adjust_transformer_in_rank` can then be provided into the constructor, and `tap_changed` can be an output parameter (just like it is for `adjust_transformer_in_rank`). That way, you can also really test per iteration

Reply:
I don't get why we would do that

Reply:
Addressed most of these points - as discussed offline